M2351: Support TFM level 1

ccli8 opened the pull request:
### Description

This PR adds TFM level 1 support on NUMAKER_PFM_M2351 target. It has the following major changes compared to no TFM integration:

1.  Target names for building TFM/non-TFM code:
    - `NUMAKER_PFM_M2351_S`: Target name for building M2351 TFM secure code
    - `NUMAKER_PFM_M2351_NS`: Target name for building M2351 TFM non-secure code
    - `NUMAKER_PFM_M2351_NOPSA_S`: Target name for building M2351 non-TFM secure code
    - `NUMAKER_PFM_M2351_NOPSA_NS`: Target name for building M2351 non-TFM non-secure code
1. To disable pre-built secure image and provide custom one, remove `NU_PREBUILD_SECURE` from `target.extra_labels` in **mbed_app.json** like:
    ```json
    "NUMAKER_PFM_M2351_NS": {
            "target.extra_labels_remove"        : ["NU_PREBUILD_SECURE"],
    ```
1. Prior to TFM, the target name `NUMAKER_PFM_M2351` is used for building non-TFM both secure/non-secure code via Nuvoton samples [NuMaker-mbed-TZ-secure-example](https://github.com/OpenNuvoton/NuMaker-mbed-TZ-secure-example) and [NuMaker-mbed-TZ-nonsecure-example](https://github.com/OpenNuvoton/NuMaker-mbed-TZ-nonsecure-example). Committed to keeping supporting non-TFM target,  new target names `NUMAKER_PFM_M2351_NOPSA_S` / `NUMAKER_PFM_M2351_NOPSA_NS` are created to substitute for `NUMAKER_PFM_M2351` to build non-TFM secure/non-secure code respectively. Their naming can be negotiated. After this PR is merged, the Nuvoton samples mentioned above will also update to reflect this change.
1. Prior to TFM, there are two steps to flash secure/non-secure code: flash secure code first and then non-secure code. In this PR, as #9221, secure code and non-secure code are combined in non-secure code build process. Just flash the combined one.
1. For TFM, no support for BL2 bootloader.
1. For TFM, there are two outstanding issues [Allow NS secure call at pre-rtos stage](https://developer.trustedfirmware.org/T376) and [Allow NS secure call in interrupt-disabled context](https://developer.trustedfirmware.org/T378).  The latter is especially critical because a complete solution hasn't come out yet. This PR temporarily modifies TFM NS code **tfm_ns_lock_rtx.c** to address the interrupt-disabled scenario partially. Please skip this related code. The thorough solution needs further discussion.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

ccli8 committed:
[M2351] Make exported NSC symbols consistent across toolchains

This idea comes from TFM.

ccli8 committed:
[M2351] Remove draft uvisor code with GCC_ARM

uVisor is deprecated since mbed-os 5.10, so remove related code with it.

ccli8 committed:
[M2351] Support initializing multiple .data/.bss sections with GCC_ARM

ccli8 committed:
[M2351] Support TFM level 1

ccli8 committed:
[M2351] Enforce locked entry into TFM for platform-specific NSC calls

On M2351, some spaces like SYS/CLK are hard-wired to secure and cannot change.
To access these spaces from non-secure world, we must provide platform-specific NSC
functions. With TFM introduced, we must synchronize NSC calls into TFM to keep TFM
in sync instead of straight NSC calls. To achieve this goal, we go with the following
approach:

1. Like PSA APIs, enforce locked entry through tfm_ns_lock_dispatch().
2. Run platform-specific secure functions in default secure partition, in which SYS/CLK
   spaces have been configured to be accessible.

ccli8 committed:
[M2351] Support secure/non-secure combined build for PSA

Support secure/non-secure combined build for PSA target:
1.  In secure post-build, deliver built secure image to TARGET_NU_PREBUILD_SECURE
    directory.
2.  In non-secure post-build, merge non-secure image with secure image saved in
    TARGET_NU_PREBUILD_SECURE directory.
3.  In non-secure post-build, user can also drop secure image saved in
    TARGET_NU_PREBUILD_SECURE directory and provide its own by adding the line below
    in mbed_app.json:
    "target.extra_labels_remove": ["NU_PREBUILD_SECURE"]

ccli8 committed:
[M2351] Change default stdio-baud-rate to 115200

ccli8 committed:
[M2351] Change memory partition for PSA

1. Change secure/non-secure ROM to 240KiB/272KiB
2. Change secure/non-secure RAM to 64KiB/32KiB

0xc0170 commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_NSPE/interface/src/tfm_ns_lock_rtx.c:
```diff
@@ -87,3 +87,17 @@ bool tfm_ns_lock_get_init_state()
+         * include extra header file like RTE_Components.h. To avoid inclusion error
+         * or inclusion of more irrelevant header files, we expand osRtxMutexId()
+         * in-place to get mutex lock count straight. */
+        //os_mutex_t *mutex = osRtxMutexId(ns_lock.id);
```
remove dead code

ccli8 replied:
There's an on-going change [Interface: Refactor the NS interface](https://review.trustedfirmware.org/c/trusted-firmware-m/+/1460) in TFM repo. The change to TF-M NS interface and related code in this PR are temporary. Just ignore it and wait for TF-M NS interface to be confirmed.

0xc0170 commented on targets/TARGET_NUVOTON/TARGET_M2351/sleep.c:
```diff
@@ -38,9 +38,7 @@ void hal_sleep(void)
-    SYS_UnlockReg_S();
-    CLK_Idle_S();
-    SYS_LockReg_S();
+    nu_idle_s();
```
how is this related to TFM support? Shouldn't this be separate PR addressing sleep issue?

0xc0170 replied:
Seeing other HAL files, this is related to the TFM ?

ccli8 replied:
With TF-M integrated in, secure call overhead gets higher. Combine 3 secure calls into one for TFM build and also for non-TFM build.

0xc0170 commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/device/TARGET_TFM/cmsis_nvic_virtual.c:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2018-2020, Nuvoton Technology Corporation
```
2019 instead of 2020 (some files have this one as well).

ccli8 replied:
Not familiar with copyright claim. Which one is better, 2018-2020, 2019-2020, or 2019?

0xc0170 replied:
just use the year this was created/edit. we are not yet in 2020

0xc0170 replied:
2018-2019 would be fine here

ccli8 replied:
@0xc0170 Fixed.

ccli8 committed:
[M2351] Override NS interface by locking kernel scheduler

Lock kernel scheduler rather than mutex to guarantee serialization of NS
secure calls

ccli8 committed:
[M2351] Merge secure idle/powerdown sequences into one

Merge SYS_UnlockReg_S()/CLK_Idle_S() or CLK_PowerDown_S()/SYS_LockReg_S()
into nu_idle_s() or nu_powerdown_s() when they are available.

ccli8 committed:
[M2351] Add pre-built secure images for PSA

urutva commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_NU_PFM_M2351_NS/TARGET_NU_PREBUILD_SECURE/README.md:
```diff
@@ -0,0 +1,39 @@
+
+These images were compiled by the following command:
+
+## tfm.hex
```
@ccli8 The TF-M secure images are generated using the python script https://github.com/ARMmbed/mbed-os/blob/master/tools/psa/release.py. The script will automatically build all the secure side images and commit them to https://github.com/ARMmbed/mbed-os/pull/10959/files#diff-3d6b492f3b697230670ddbb26104bbfdR8827. For every release of mbed-os, `release.py` is run for all PSA targets and the secure binaries are checked in to relevant NS folders.

Can you please update the README.md and also regenerate the binaries using `release.py`.

ccli8 replied:
Fixed

urutva commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_TFM/nu_tfm_ns_lock_sup.c:
```diff
@@ -0,0 +1,76 @@
+ *    SYS/CLK spaces must have configured to be accessible.
+ */
+
+int32_t nu_tfm_ns_lock_dispatch(veneer_fn fn,
```
What is the need for this Nuvoton specific API and when will it be called?
Currently, https://github.com/ARMmbed/mbed-os/blob/4c04771c5d3a8d730d41d7fdbc9859ffe7c12148/components/TARGET_PSA/TARGET_TFM/COMPONENT_NSPE/interface/src/tfm_psa_ns_api.c uses `tfm_ns_lock_dispatch` for dispatching secure calls to TF-M.

ccli8 replied:
`tfm_ns_lock_dispatch()` doesn't support secure call in pre-rtos or interrupt-disabled context, which is required to call platform-specific secure functions on M2351. In updated TF-M repo, `tfm_ns_lock_dispatch()` is renamed to `tfm_ns_interface_dispatch()` and gets weak https://github.com/ARMmbed/mbed-os/pull/10959#issuecomment-514446186. The `nu_tfm_ns_lock_dispatch(...)` is temporary for development. After new TF-M NS interface is integrated into mbed-os, `tfm_ns_interface_dispatch()` will be overridden in M2351 directory to try supporting secure call in pre-rtos or interrupt-disabled context.

ccli8 replied:
@devran01 Remove `nu_tfm_ns_lock_dispatch(...)`. Make `tfm_ns_lock_dispatch()` weak and override it to support secure call in unusual conditions like pre-rtos and interrupt-disabled.

urutva commented on targets/targets.json:
```diff
@@ -8778,6 +8777,65 @@
+    "NU_PFM_M2351_NS": {
+        "inherits": ["NSPE_Target", "NU_PFM_M2351_CM"],
+        "core": "Cortex-M23-NS",
+        "tfm.level": 1,
```
@ccli8  TFM_LVL is already defined in https://github.com/ARMmbed/mbed-os/blob/master/components/TARGET_PSA/TARGET_TFM/mbed_lib.json#L7. What is the need for this?

ccli8 replied:
Not sure if the default `TFM_LVL` will change. Set it explicitly for safe.

urutva commented on targets/targets.json:
```diff
@@ -8880,7 +8938,7 @@
         ],
         "release_versions": ["5"],
-        "extra_labels": ["Cypress", "PSOC6", "MXCRYPTO"],
+        "extra_labels": ["Cypress", "PSOC6"],
```
@ccli8 @0xc0170 Seems like some issue with rebasing. Why is this PR trying to update other targets? Please check.

ccli8 replied:
Sorry. I fixed them. It may result after several rebase operations.

VeijoPesonen commented on targets/targets.json:
```diff
@@ -8778,6 +8777,65 @@
+        "post_binary_hook": {"function": "M2351Code.merge_secure"},
+        "secure_image_filename": "tfm.hex",
+        "overrides": {
+            "secure-rom-start"      : "0x0",
```
Why the non-secure side needs to know about the fact from where does the secure side start?

ccli8 replied:
Yes, non-secure side needn't. However, in rare case like `mbed_hal-spm` test, secure memory is intentionally accessed to trigger [security fault](https://github.com/ARMmbed/mbed-os/blob/eeb033f8148a87659f3c084ec0b69e59ba9ac78e/TESTS/mbed_hal/spm/main.cpp#L102), though this test is filtered out because SecureFault cannot remap.

VeijoPesonen commented on targets/targets.json:
```diff
@@ -8778,6 +8777,65 @@
+            "secure-rom-size"       : "0x3C000",
+            "secure-ram-start"      : "0x20000000",
+            "secure-ram-size"       : "0x10000",
+            "non-secure-rom-start"  : "0x1003C000",
+            "non-secure-rom-size"   : "0x44000",
+            "non-secure-ram-start"  : "0x30010000",
+            "non-secure-ram-size"   : "0x8000"
```
Maybe drop the `non-secure-`-prefix?

ccli8 replied:
No. They are both defined in [PSA_Target](https://github.com/ARMmbed/mbed-os/blob/eeb033f8148a87659f3c084ec0b69e59ba9ac78e/targets/targets.json#L61).

VeijoPesonen commented on targets/targets.json:
```diff
@@ -8778,6 +8777,65 @@
         "mbed_ram_size"         : "0x8000"
     },
+    "NU_PFM_M2351_NS": {
+        "inherits": ["NSPE_Target", "NU_PFM_M2351_CM"],
```
`NU_PFM_M2351_CM`? is it like that there is common between secure- and non-secure-side? If it is so why it's not named as `NU_PFM_M2351`? That's how it's done with `LPC55S69` and having things done consistently is always good.

ccli8 replied:
`NU_PFM_M2351` is reserved for special target like **No TrustZone** target, though it is not supported currently.

VeijoPesonen commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_NSPE/interface/src/tfm_ns_lock_rtx.c:
```diff
@@ -42,6 +42,7 @@ static const osMutexAttr_t ns_lock_attrib = {
 /**
  * \brief NS world, NS lock based dispatcher
  */
+__attribute__((weak))
```
You should use MBED_WEAK instead.

```
/** MBED_WEAK
 *  Mark a function as being weak.
 *
 *  @note
 *  Functions should only be marked as weak in the source file. The header file
 *  should contain a regular function declaration to insure the function is emitted.
 *  A function marked weak will not be emitted if an alternative non-weak
 *  implementation is defined.
 *
 *  @note
 *  Weak functions are not friendly to making code re-usable, as they can only
 *  be overridden once (and if they are multiply overridden the linker will emit
 *  no warning). You should not normally use weak symbols as part of the API to
 *  re-usable modules.
 *
 *  @code
 *  #include "mbed_toolchain.h"
 *
 *  MBED_WEAK void foo() {
 *      // a weak implementation of foo that can be overriden by a definition
 *      // without  __weak
 *  }
 *  @endcode
 */
#ifndef MBED_WEAK
#if defined(__ICCARM__)
#define MBED_WEAK __weak
#elif defined(__MINGW32__)
#define MBED_WEAK
#else
#define MBED_WEAK __attribute__((weak))
#endif
#endif
```
- platform/mbed_toolchain.h

ccli8 replied:
The change to this file is temporary for my porting. Its final modification should come from TF-M repo (see https://github.com/ARMmbed/mbed-os/pull/10959#issuecomment-543799527).

ccli8 replied:
@VeijoPesonen I still use `__attribute__((weak))` because all toolchains support it. This is also what https://review.trustedfirmware.org/c/trusted-firmware-m/+/1460 is adopting.

VeijoPesonen commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_S/TARGET_TFM/target_cfg.cpp:
```diff
@@ -0,0 +1,106 @@
+
+extern "C" void enable_fault_handlers(void)
+{
+    /* M2351 doesn't implement Main Extension, so these faults are not supported. */
```
Please remove code which isn't used.

ccli8 replied:
Removed

VeijoPesonen commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_S/TARGET_TFM/target_cfg.h:
```diff
@@ -0,0 +1,50 @@
+{
+    uint32_t periph_start;
+    uint32_t periph_limit;
+    /* TODO */
```
What needs to be done?

ccli8 replied:
Removed it

VeijoPesonen commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_S/TARGET_TFM/tfm_peripherals_def.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef __TFM_PERIPHERALS_DEF_H__
+#define __TFM_PERIPHERALS_DEF_H__
+
+/* TODO */
```
What needs to be done?

ccli8 replied:
Removed it

VeijoPesonen commented on targets/TARGET_NUVOTON/TARGET_M2351/device/partition_M2351_mem.h:
```diff
@@ -108,4 +150,124 @@
+ * 1. Requested by SAU, NSC area must start at 32 byte-aligned boundary.
+ * 2. By IDAU, 0~0x4000 is secure. NSC can only locate in 0x4000~0x10000000.
+ * 3. Greentea flash IAP uses last 2 sectors for its test. Avoid this range.
+ * 4. Greentea NVSTORE uses last 2 sectors or 4 KiB x 2 for its test. Avoid this range.
```
@SeppoTakalo for your information.

VeijoPesonen reviewed (approved):
From those parts I do understand things look good to me. Only small remarks which don't require any actions codewise.

0xc0170 commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_S/device/TOOLCHAIN_IAR/M2351.icf:
```diff
@@ -0,0 +1,73 @@
+define block Image$$ER_IROM_NSC with alignment = 32 { readonly section Veneer$$CMSE };
+
+define block CSTACK     with alignment = 8, size = __ICFEDIT_size_cstack__   { };
+define block HEAP       with alignment = 8, size = __ICFEDIT_size_heap__     { };
```
should this use dynamic heap as well?

ccli8 replied:
@0xc0170 Added dynamic heap configuration on IAR, though, M2351 doesn't support IAR yet.

ccli8 committed:
TFM: Change NS lock functions to weak

Default NS lock functions use mutex to guarantee serialization into secure world.
But this mechanism doesn't support pre-rtos and interrupt-disabled.
This commit changes these functions to weak so that platform can override them to provide another mechanism.
This approach is just what TF-M repository is adopting.

https://review.trustedfirmware.org/c/trusted-firmware-m/+/1460

madchutney reviewed (dismissed):
`NUMAKER_PFM_M2351_NOPSA_S` and `NUMAKER_PFM_M2351_NOPSA_NS` are longer than the 20 character limit which is currently enforced by the online schema.

I believe we need to adopt another system for Mbed 6 and not combine the type of build with the board name.
